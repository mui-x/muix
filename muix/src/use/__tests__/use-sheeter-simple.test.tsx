/** @jsx platform.createElement */

import { platform } from "reactxx-sheeter"

import { TSheeter, TWithStyles, TComponents } from 'reactxx-typings'
import { useSheeter } from "reactxx-use"

import { initPlatform, mount } from "./init-platform"

interface Shape extends TSheeter.ShapeAncestor {
  rulesets: {
    root: 'Text'
  },
  props: {
    p1: string
  },
  theme: {
    primaryColor: string
    p1Prop: string
  }
}

describe("USE SHEETER", () => {

  const doTest = (isWeb: boolean) => {
    beforeEach(() => {
      initPlatform(isWeb, { dataTraceFlag: 'short' })
      platform.getDefaultTheme = () => ({
        primaryColor: 'gray',
        p1Prop: 'themeProp'
      })
    })

    const compCreator = (
      config: TWithStyles.ComponentConfig<Shape>,
      configOverride: TWithStyles.ComponentConfigOverride<Shape>,
      displayName?: 'Comp'
    ) => {
      const res: TComponents.SFC<Shape> = props => {
        try {
          const { toClassNames, propsCode, classes, classNameX, styleX
          } = useSheeter(props, config, displayName, configOverride)
          
          const root = toClassNames(classes.root, classNameX)
          return <div {...propsCode} classNameX={root} styleX={styleX as any} />
        } catch {
          return <div>ERROR</div>
        }
      }
      res.displayName = displayName
      res['$c$'] = true
      return res
    }


    it('01: config NULL', () => {
      const Comp = compCreator(null, null)
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('02: sheet missing', () => {
      const Comp = compCreator({}, null)
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('03: just div', () => {
      const Comp = compCreator({ defaultSheet: { root: {} } }, null)
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('04: with sheet', () => {
      const Comp = compCreator({ defaultSheet: { root: { color: 'red' } } }, null)
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('05: with sheet override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('06: with classes override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={{ root: { color: 'blue' } }} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('07: with classNameX override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={{ root: { color: 'blue' } }} classNameX={{ color: 'yellow' } as any} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('08: with styleX override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={{ root: { color: 'blue' } }} classNameX={{ color: 'yellow' } as any} styleX={{ color: 'maroon' } as any} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('09: theme, with sheet', () => {
      const Comp = compCreator({ defaultSheet: theme => ({ root: { color: theme.primaryColor } }) }, null)
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('10: theme, with sheet override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: theme => ({ root: { color: theme.primaryColor } }) }
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('11: theme, with classes override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={theme => ({ root: { color: theme.primaryColor } })} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('12: theme, with classNameX override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={{ root: { color: 'blue' } }} classNameX={theme => ({ color: theme.primaryColor }) as any} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('13: theme, with styleX override', () => {
      const Comp = compCreator(
        { defaultSheet: { root: { color: 'red' } } },
        { overrideSheet: { root: { color: 'green' } } }
      )
      const wrapper = mount(<Comp classes={{ root: { color: 'blue' } }} classNameX={{ color: 'yellow' } as any} styleX={theme => ({ color: theme.primaryColor }) as any} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })


    it('14: default props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp' }
        },
        null
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('15: override props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp' }
        },
        { overrideProps: { p1: 'overrideProp' } }
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('16: props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp' }
        },
        { overrideProps: { p1: 'overrideProp' } }
      )
      const wrapper = mount(<Comp p1='componentProp' />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('17: theme, default props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp', themedProps: theme => ({ p1: theme.p1Prop }) }
        },
        null
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('18: theme, override props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp' }
        },
        { overrideProps: { p1: 'overrideProp', themedProps: theme => ({ p1: theme.p1Prop }) } }
      )
      const wrapper = mount(<Comp />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

    it('19: theme, props', () => {
      const Comp = compCreator(
        {
          defaultSheet: { root: {} },
          defaultProps: { p1: 'defaultProp' }
        },
        { overrideProps: { p1: 'overrideProp' } }
      )
      const wrapper = mount(<Comp p1='componentProp' themedProps={theme => ({ p1: theme.p1Prop })} />)
      expect(wrapper.container).toMatchSnapshot()
      wrapper.unmount()
    })

  }

  describe("## NATIVE ##", () => doTest(false))
  describe("## WEB ##", () => doTest(true))

})

