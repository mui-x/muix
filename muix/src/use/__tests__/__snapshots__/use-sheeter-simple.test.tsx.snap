// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USE SHEETER ## NATIVE ## 01: config NULL 1`] = `
<div>
  <div>
    ERROR
  </div>
</div>
`;

exports[`USE SHEETER ## NATIVE ## 02: sheet missing 1`] = `
<div>
  <div>
    ERROR
  </div>
</div>
`;

exports[`USE SHEETER ## NATIVE ## 03: just div 1`] = `
<div>
  <div />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 04: with sheet 1`] = `
<div>
  <div
    data-trace=""
    style="color: red;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 05: with sheet override 1`] = `
<div>
  <div
    data-trace=""
    style="color: green;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 06: with classes override 1`] = `
<div>
  <div
    data-trace=""
    style="color: blue;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 07: with classNameX override 1`] = `
<div>
  <div
    data-trace=""
    style="color: yellow;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 08: with styleX override 1`] = `
<div>
  <div
    data-trace=""
    style="color: maroon;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 09: theme, with sheet 1`] = `
<div>
  <div
    data-trace=""
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 10: theme, with sheet override 1`] = `
<div>
  <div
    data-trace=""
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 11: theme, with classes override 1`] = `
<div>
  <div
    data-trace=""
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 12: theme, with classNameX override 1`] = `
<div>
  <div
    data-trace=""
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 13: theme, with styleX override 1`] = `
<div>
  <div
    data-trace=""
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 14: default props 1`] = `
<div>
  <div
    p1="defaultProp"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 15: override props 1`] = `
<div>
  <div
    p1="overrideProp"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 16: props 1`] = `
<div>
  <div
    p1="componentProp"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 17: theme, default props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 18: theme, override props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;

exports[`USE SHEETER ## NATIVE ## 19: theme, props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 01: config NULL 1`] = `
<div>
  <div>
    ERROR
  </div>
</div>
`;

exports[`USE SHEETER ## WEB ## 02: sheet missing 1`] = `
<div>
  <div>
    ERROR
  </div>
</div>
`;

exports[`USE SHEETER ## WEB ## 03: just div 1`] = `
<div>
  <div />
</div>
`;

exports[`USE SHEETER ## WEB ## 04: with sheet 1`] = `
<div>
  <div
    class="a"
    data-trace="
.a { color:red }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 05: with sheet override 1`] = `
<div>
  <div
    class="b"
    data-trace="
.b { color:green }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 06: with classes override 1`] = `
<div>
  <div
    class="c"
    data-trace="
.c { color:blue }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 07: with classNameX override 1`] = `
<div>
  <div
    class="d"
    data-trace="
.d { color:yellow }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 08: with styleX override 1`] = `
<div>
  <div
    class="d"
    data-trace="
.d { color:yellow }"
    style="color: maroon;"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 09: theme, with sheet 1`] = `
<div>
  <div
    class="a"
    data-trace="
.a { color:gray }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 10: theme, with sheet override 1`] = `
<div>
  <div
    class="b"
    data-trace="
.b { color:gray }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 11: theme, with classes override 1`] = `
<div>
  <div
    class="c"
    data-trace="
.c { color:gray }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 12: theme, with classNameX override 1`] = `
<div>
  <div
    class="d"
    data-trace="
.d { color:gray }"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 13: theme, with styleX override 1`] = `
<div>
  <div
    class="d"
    data-trace="
.d { color:yellow }"
    style="color: gray;"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 14: default props 1`] = `
<div>
  <div
    p1="defaultProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 15: override props 1`] = `
<div>
  <div
    p1="overrideProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 16: props 1`] = `
<div>
  <div
    p1="componentProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 17: theme, default props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 18: theme, override props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;

exports[`USE SHEETER ## WEB ## 19: theme, props 1`] = `
<div>
  <div
    p1="themeProp"
  />
</div>
`;
